File: ui/src/api/hooks/push.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api, type PushDeliveryReport, type PushStatus } from "../openapiClient";
import { unwrap } from "../openapiClient";
import { isBackendAvailable } from "../../utils/runtime";

/** VAPID public key + how many devices this user has registered. */
export function usePushStatus() {
  return useQuery<PushStatus>({
    queryKey: ["push-status"],
    queryFn: async () => {
      return unwrap(api.getPushStatus());
    },
    // The key is generated once and never rotates; only device_count moves, and
    // only in response to actions taken on this screen.
    staleTime: 5 * 60_000,
    enabled: isBackendAvailable(),
  });
}

export function useSavePushSubscription() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (sub: { endpoint: string; p256dh: string; auth: string; label?: string }) => {
      return unwrap(api.savePushSubscription(
        sub.endpoint,
        sub.p256dh,
        sub.auth,
        sub.label ?? null
      ));
    },
    onSuccess: () => void qc.invalidateQueries({ queryKey: ["push-status"] }),
  });
}

export function useDeletePushSubscription() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (endpoint: string) => {
      return unwrap(api.deletePushSubscription(endpoint));
    },
    onSuccess: () => void qc.invalidateQueries({ queryKey: ["push-status"] }),
  });
}

/** Send a notification to this user's devices so they can confirm it works. */
export function useSendTestPush() {
  return useMutation<PushDeliveryReport>({
    mutationFn: async () => {
      return unwrap(api.sendTestPush());
    },
  });
}

File: crates/finsight-app/src/commands/push.rs
use crate::{error::AppResult, AppState};

pub use finsight_api::commands::push::{PushDeliveryReport, PushDevice, PushStatus};

#[tauri::command]
#[specta::specta]
pub async fn get_push_status(state: tauri::State<'_, AppState>) -> AppResult<PushStatus> {
    finsight_api::commands::push::get_push_status(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn save_push_subscription(
    state: tauri::State<'_, AppState>,
    endpoint: String,
    p256dh: String,
    auth: String,
    label: Option<String>,
) -> AppResult<()> {
    finsight_api::commands::push::save_push_subscription(&state.api, endpoint, p256dh, auth, label)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn delete_push_subscription(
    state: tauri::State<'_, AppState>,
    endpoint: String,
) -> AppResult<bool> {
    finsight_api::commands::push::delete_push_subscription(&state.api, endpoint).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_push_devices(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<PushDevice>> {
    finsight_api::commands::push::list_push_devices(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn send_test_push(
    state: tauri::State<'_, AppState>,
) -> AppResult<PushDeliveryReport> {
    finsight_api::commands::push::send_test_push(&state.api).await
}
